python/dgl/distributed/kvstore.py

```diff
@@ -0,0 +1,1008 @@
+                             GetPartShapeResponse)
+        rpc.register_service(SEND_META_TO_BACKUP,
+                             SendMetaToBackupRequest,
+                             SendMetaToBackupResponse)
+        # Store the tensor data with specified data name
+        self._data_store = {}
+        # Store the partition information with specified data name
+        self._part_policy = {}
+        # Basic information
+        self._server_id = server_id
+        self._server_namebook = rpc.read_ip_config(ip_config)
+        # TODO(chao): machine_id can be removed if we use new shared-memory API
+        self._machine_id = self._server_namebook[server_id][0]
+        self._group_count = self._server_namebook[server_id][3]
+        # We assume partition_id is equal to machine_id
+        self._part_id = self._machine_id
+        self._num_clients = num_clients
+        self._barrier_count = 0
+        # push and pull handler
+        self._push_handler = default_push_handler
+        self._pull_handler = default_pull_handler
+
+    @property
+    def server_id(self):
+        return self._server_id
+
+    @property
+    def barrier_count(self):
+        return self._barrier_count
+
+    def incr_barrier_count(self):
+        self._barrier_count +=1
+
+    def barrier_zero(self):
+        self._barrier_count = 0
+
+    @property
+    def num_clients(self):
+        return self._num_clients
+     
+    @property
+    def data_store(self):
+        return self._data_store
+
+    @property
+    def part_policy(self):
+        return self._part_policy
+     
+    @property
+    def push_handler(self):
+        return self._push_handler
+
+    @property
+    def pull_handler(self):
+        return self._pull_handler
+
+    @pull_handler.setter
+    def pull_handler(self, pull_handler):
+        self._pull_handler = pull_handler
+
+    @push_handler.setter
+    def push_handler(self, push_handler):
+        self._push_handler = push_handler
+
+    def is_backup_server(self):
+        """Return True if current server is a backup server.
+        """
+        if self._server_id % self._group_count == 0:
+            return False
+        return True
+
+    def init_data(self, name, data_tensor=None):
+        """Init data tensor on kvserver.
+
+        Parameters
+        ----------
+        name : str
+            data name
+        data_tensor : tensor
+            If the data_tensor is None, KVServer will
+            read shared-memory when client invoking get_shared_data().
+        """
+        assert len(name) > 0, 'name cannot be empty.'
+        if data_tensor is not None: # Create shared-tensor
+            data_type = F.reverse_data_type_dict[F.dtype(data_tensor)]
+            shared_data = empty_shared_mem(name+'-kvdata-', True, data_tensor.shape, data_type)
+            dlpack = shared_data.to_dlpack()
+            self._data_store[name] = F.zerocopy_from_dlpack(dlpack)
+            self._data_store[name][:] = data_tensor[:]
```

Comment:
it doesn't do anything if `data_tensor` is None?

Reply:
I change the logic here. If data_tensor = None, backup_server will do nothing. The backup-server read the shared-tensor only when client invoke get_shared() API.

Reply:
can you put comments here?